[gfx906] Fix unified attention tiling and WNA16 fallback defaults

tomylin890 opened the pull request:
## Summary
- **Cause → Fix → Impact**: v0.11.0 routes gfx906 through the unified-attention gather loop whenever sliding window is disabled, which regresses prefill throughput; realign the unified-attention tile size with the full block so gfx906 reuses the fast path and restores performance.
- **Cause → Fix → Impact**: AWQ/GPTQ MoE kernels assert on `hidden_size % BLOCK_SIZE_K == 0` when the WNA16 fallback picks a non-divisor; choose the fallback BLOCK_SIZE_K so it always divides the hidden size, allowing models such as Qwen3-VL to load without a tuned config on gfx906.

## Testing
- [x] `vllm serve /model/Qwen3-30B-A3B-Thinking-2507-GPTQ-Int8 --tensor-parallel-size 4 --max-num-batched-tokens 6144 --max-num-seqs 48 --gpu-memory-utilization 0.95 --max-model-len 262144` (TP=4, warm start, 10 prompt smoke test)
- Note: gfx906-only change; other GPUs are unaffected.

## Performance (gfx906)
- **Delta**: Prefill throughput improves by ~30% at ~10K tokens; decode speed remains within measurement noise.
- **Environment**: 4 × gfx906 (ROCm 6.3.3—update to the exact driver version you used).

comment: gfx906 vllm v0.11.0 origin, model: /model/Qwen3-30B-A3B-Thinking-2507-GPTQ-Int8

| prompt length (tokens) | prefill time (ms) | prefill speed (tokens/s) | output length (tokens) | output time (ms) | output speed (tokens/s) |
|------------------------|-------------------|---------------------------|------------------------|------------------|--------------------------|
| 512 | 398.00 | 1286.43 | 128 | 1858.00 | 68.89 |
| 5632 | 2590.00 | 2174.52 | 128 | 2481.00 | 51.59 |
| 10752 | 7393.00 | 1454.35 | 128 | 2977.00 | 43.00 |
| 15872 | 14182.00 | 1119.17 | 128 | 3438.00 | 37.23 |
| 20992 | 23705.00 | 885.55 | 128 | 3917.00 | 32.68 |
| 26112 | 35139.00 | 743.11 | 128 | 4356.00 | 29.38 |
| 31232 | 49575.00 | 629.99 | 128 | 4812.00 | 26.60 |
| 36352 | 65748.00 | 552.90 | 128 | 5256.00 | 24.35 |
| 41472 | 85076.00 | 487.47 | 128 | 5707.00 | 22.43 |

comment: gfx906 vllm v0.11.0 with this PR, model: /model/Qwen3-30B-A3B-Thinking-2507-GPTQ-Int8

| prompt length (tokens) | prefill time (ms) | prefill speed (tokens/s) | output length (tokens) | output time (ms) | output speed (tokens/s) |
|------------------------|-------------------|---------------------------|------------------------|------------------|--------------------------|
| 512 | 171.00 | 2994.15 | 128 | 1866.00 | 68.60 |
| 5632 | 1940.00 | 2903.09 | 128 | 2478.00 | 51.65 |
| 10752 | 5058.00 | 2125.74 | 128 | 2984.00 | 42.90 |
| 15872 | 9400.00 | 1688.51 | 128 | 3465.00 | 36.94 |
| 20992 | 15123.00 | 1388.08 | 128 | 3914.00 | 32.70 |
| 26112 | 22132.00 | 1179.83 | 128 | 4380.00 | 29.22 |
| 31232 | 30488.00 | 1024.40 | 128 | 4824.00 | 26.53 |
| 36352 | 39723.00 | 915.14 | 128 | 5269.00 | 24.29 |
| 41472 | 50590.00 | 819.77 | 128 | 5733.00 | 22.33 |

tomylin890 committed:
[gfx906] Reduce unified attention tiling overhead

- Keep TILE_SIZE equal to block size when window_size[0] < 0 so full-attention models avoid the extra gather loop introduced in v0.11.0.
- Preserves sliding-window handling but restores near-v0.10.x prefill performance for Qwen3MoE on MI50/MI60.

tomylin890 committed:
[gfx906] Fix WNA16 fallback block config

- When no tuned config file exists, choose a BLOCK_SIZE_K that divides hidden size so moe_wna16_gemm stops asserting.
- Applies to AWQ/GPTQ MoE models like Qwen3-VL on MI50 where we fall back to default config generation.

dagzatlelot commented:
Thanks for the work, seems like a great performance improvement.
I have some question regarding your comment
> allowing models such as Qwen3-VL to load without a tuned config on gfx906

Are you actually able to use Qwen3-VL? I thought Qwen2.5-VL/Qwen3-VL fallback to torch's SDPA implementation, which has issues with memory allocation on gfx906. At least that's the case for me, even with your PR.

tomylin890 commented:
> Thanks for the work, seems like a great performance improvement. I have some question regarding your comment
> 
> > allowing models such as Qwen3-VL to load without a tuned config on gfx906
> 
> Are you actually able to use Qwen3-VL? I thought Qwen2.5-VL/Qwen3-VL fallback to torch's SDPA implementation, which has issues with memory allocation on gfx906. At least that's the case for me, even with your PR.

In this PR, I used the Qwen3-VL-30B-A3B-Thinking-AWQ-4bit model, the model URL is: https://huggingface.co/cpatonn/Qwen3-VL-30B-A3B-Thinking-AWQ-4bit.

The model can be loaded and used for visual inference normally. I have tested it on several photos of handwritten characters, and they can be recognized correctly, even Chinese characters.

This may be a special case, but at least I have successfully used it with this version of vllm.

However, when I was doing the full attention performance fix, it seems that I broke the loading of the Qwen3Next series model. I am working on a further fix, which should be integrated into this PR in the next few hours.

This could also be a trial patch, at least to let us know that the gfx906 can run visual models and perform well (pp40960 ~ 700tks with tp4).

dagzatlelot commented:
I've rebuilt your branch and this time it actually worked, so there was some issue on my side. It runs pretty well.
Thanks again for your contribution.

tomylin890 commented:
Update1:

Fixed the Qwen3Next pathway, and the performance after warmup:

comment: gfx906 vllm v0.11.0 with this PR, model: /model/Qwen3-Next-80B-A3B-Thinking-AWQ-8bit

| prompt length (tokens) | prefill time (ms) | prefill speed (tokens/s) | output length (tokens) | output time (ms) | output speed (tokens/s) |
|---------------------|-------------------|-------------------------|-------------------|---------------|---------------------|
| 512 | 325.00 | 1575.38 | 128 | 4072.00 | 31.43 |
| 5632 | 2712.00 | 2076.70 | 128 | 4610.00 | 27.77 |
| 10752 | 6534.00 | 1645.55 | 128 | 5085.00 | 25.17 |
| 15872 | 10946.00 | 1450.03 | 128 | 5602.00 | 22.85 |
| 20992 | 16861.00 | 1245.00 | 128 | 6119.00 | 20.92 |
| 26112 | 23315.00 | 1119.97 | 128 | 6638.00 | 19.28 |
| 31232 | 31329.00 | 996.90 | 128 | 7167.00 | 17.86 |
| 36352 | 39819.00 | 912.93 | 128 | 7685.00 | 16.66 |
| 41472 | 49987.00 | 829.66 | 128 | 8216.00 | 15.58 |

And remand the Qwen3MoE performance:

comment: gfx906 vllm v0.11.0 with this PR, model: /model/Qwen3-30B-A3B-Thinking-2507-GPTQ-Int8 

| prompt length (tokens) | prefill time (ms) | prefill speed (tokens/s) | output length (tokens) | output time (ms) | output speed (tokens/s) |
|---------------------|-------------------|-------------------------|-------------------|---------------|---------------------|
| 512 | 170.00 | 3011.76 | 128 | 1880.00 | 68.09 |
| 5632 | 1938.00 | 2906.09 | 128 | 2515.00 | 50.89 |
| 10752 | 5054.00 | 2127.42 | 128 | 2982.00 | 42.92 |
| 15872 | 9410.00 | 1686.72 | 128 | 3450.00 | 37.10 |
| 20992 | 15135.00 | 1386.98 | 128 | 3913.00 | 32.71 |
| 26112 | 22145.00 | 1179.14 | 128 | 4374.00 | 29.26 |
| 31232 | 30468.00 | 1025.08 | 128 | 4823.00 | 26.54 |
| 36352 | 39741.00 | 914.72 | 128 | 5263.00 | 24.32 |
| 41472 | 50588.00 | 819.80 | 128 | 5729.00 | 22.34 |

Also tested the Qwen3VL performance:

comment: gfx906 vllm v0.11.0 with this PR, model: /model/Qwen3-VL-30B-A3B-Thinking-AWQ-4bit

| prompt length (tokens) | prefill time (ms) | prefill speed (tokens/s) | output length (tokens) | output time (ms) | output speed (tokens/s) |
|---------------------|-------------------|-------------------------|-------------------|---------------|---------------------|
| 512 | 320.00 | 1600.00 | 128 | 2039.00 | 62.78 |
| 5632 | 2685.00 | 2097.58 | 128 | 2579.00 | 49.63 |
| 10752 | 6442.00 | 1669.05 | 128 | 3062.00 | 41.80 |
| 15872 | 11449.00 | 1386.32 | 128 | 3525.00 | 36.31 |
| 20992 | 17821.00 | 1177.94 | 128 | 3962.00 | 32.31 |
| 26112 | 25468.00 | 1025.29 | 128 | 4418.00 | 28.97 |
| 31232 | 34472.00 | 906.01 | 128 | 4869.00 | 26.29 |
| 36352 | 44325.00 | 820.12 | 128 | 5343.00 | 23.96 |
| 41472 | 55814.00 | 743.04 | 128 | 5820.00 | 21.99 |

If anyone wants to try it out, they can pull my v0.11.0 branch, where these targeted adaptations will be implemented before the PR is merged.

p.s. All of these tests were with tp=4.

tomylin890 committed:
[gfx906] Adjust Triton attention tile sizing

- keep v0.11.0's sliding-window behavior intact while replacing the hardcoded 16/32 tile sizes with a shared-memory-aware picker
- select the closest power-of-two tile that fits both the attention block size and the ROCm shared memory budget, falling back to safe minimums for fp8 / non-fp8 paths
- restore full-attention MoE throughput on gfx906 and keep hybrid / sliding-window models stable on the same kernel

unverbraucht commented:
Hi @tomylin890 , I tried your changes alongside the [warp-reduce changes for triton-gfx906](https://github.com/nlzy/triton-gfx906/pull/2). I see a nice speed improvement and have not noticed any issues.

Here's vllm-bench with Qwen3-4b-Instruct-2507 AWQ from v0.11.0 (without your changes):
```

                             Output tokens per second
  400 +---------------------------------------------------------------------+
      |    * *   *  *   **   * *  **  **   **  **   *  *   *  * *  *        |
  350 |    *  * *    * *  * *  * * * *  * * * *  * ** * * ** *   * *        |
      |   *   * *    * *  * *   **  **  * *  **  * * ** * * **   ** *       |
      |   *    *      *    *    *   *    *   *    *  *   *  *     * *       |
  300 | ***    *      *    *    *        *        *  *   *  *     * *       |
      | *                                                           *       |
  250 | *                                                           *       |
      | *                                                            *      |
  200 |*                                                             *      |
      |*                                                             *      |
      |*                                                             *      |
  150 |*                                                             *      |
      |*                                                              *     |
  100 |*                                                              *     |
      |*                                                              *     |
      |                                                               *     |
   50 |                                                                *    |
      |                                                                *    |
    0 +---------------------------------------------------------------------+
      0        5        10      15       20       25       30      35       40

============ Serving Benchmark Result ============
Successful requests:                     100       
Benchmark duration (s):                  36.61     
Total input tokens:                      102202    
Total generated tokens:                  12673     
Request throughput (req/s):              2.73      
Output token throughput (tok/s):         346.18    
Peak output token throughput (tok/s):    396.00    
Peak concurrent requests:                100.00    
Total Token throughput (tok/s):          3137.92   
---------------Time to First Token----------------
Mean TTFT (ms):                          1424.75
Median TTFT (ms):                        1448.46
P99 TTFT (ms):                           1474.46
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          276.72
Median TPOT (ms):                        276.71
P99 TPOT (ms):                           276.94
---------------Inter-token Latency----------------
Mean ITL (ms):                           276.78
Median ITL (ms):                         276.68
P99 ITL (ms):                            288.29
==================================================
```

And here's with this PR:
```
                             Output tokens per second
  500 +---------------------------------------------------------------------+
      |   * *  **   *  *   **  * *  **     **   **  * *    * *  **     *    |
  450 |   * * *  * *    * * *  * * *  *   *  * * *  * *    * * *  *   * *   |
      |  *   **  * *    * *  **   **  *   *  * *  **   *  *   **  *   * *   |
  400 | *    *    *      *   *    *    ***    *   *    ***    *    ***   *  |
      | *                                                                *  |
  350 | *                                                                *  |
      | *                                                                 * |
  300 |*                                                                  * |
  250 |*                                                                  * |
      |*                                                                  * |
  200 |*                                                                   *|
  350 | *                                                                *  |
      | *                                                                 * |
  300 |*                                                                  * |
  250 |*                                                                  * |
      |*                                                                  * |
  200 |*                                                                   *|
      |*                                                                   *|
  150 |*                                                                   *|
      |*                                                                   *|
  100 |*                                                                   *|
      |                                                                    *|
   50 |                                                                     |
      |                                                                     |
    0 +---------------------------------------------------------------------+
      0           5          10          15          20         25          30

============ Serving Benchmark Result ============
Successful requests:                     100       
Benchmark duration (s):                  29.72     
Total input tokens:                      102202    
Total generated tokens:                  12673     
Request throughput (req/s):              3.36      
Output token throughput (tok/s):         426.39    
Peak output token throughput (tok/s):    495.00    
Peak concurrent requests:                100.00    
Total Token throughput (tok/s):          3864.99   
---------------Time to First Token----------------
Mean TTFT (ms):                          1259.64   
Median TTFT (ms):                        1278.30   
P99 TTFT (ms):                           1464.98   
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          223.79    
Median TPOT (ms):                        223.77    
P99 TPOT (ms):                           223.99    
---------------Inter-token Latency----------------
Mean ITL (ms):                           223.83    
Median ITL (ms):                         224.13    
P99 ITL (ms):                            233.79    
==================================================
```

nlzy commented:
Really sorry for the late reply!

Thanks so much for your code, this PR brings a significant performance boost, and I'm very happy to merge it!

I've merged it into `gfx906/main` and `gfx906/v0.11.0` branches, so this optimization will be included in `v0.11.0.update1` and future versions.
